client/state/sites/domains/selectors.js

```diff
@@ -74,17 +68,24 @@ export const getDecoratedSiteDomains = treeSelect(
 			return null;
 		}
 
-		return domains.map( domain => ( {
-			...domain,
-			autoRenewalMoment: domain.autoRenewalDate ? moment( domain.autoRenewalDate ) : null,
-			registrationMoment: domain.registrationDate ? moment( domain.registrationDate ) : null,
-			expirationMoment: domain.expiry ? moment( domain.expiry ) : null,
-			transferAwayEligibleAtMoment: domain.transferAwayEligibleAt
-				? moment( domain.transferAwayEligibleAt )
-				: null,
-			transferEndDateMoment: domain.transferStartDate
-				? moment( domain.transferStartDate ).add( 7, 'days' )
-				: null,
-		} ) );
+		return domains.map( domain => {
+			let transferEndDate;
+
+			if ( domain.transferStartDate ) {
+				transferEndDate = new Date( domain.transferStartDate );
+				transferEndDate.setDate( transferEndDate.getDate() + 7 ); // Add 7 days.
+			}
+
+			return {
+				...domain,
+				autoRenewalDate: domain.autoRenewalDate ? new Date( domain.autoRenewalDate ) : null,
```

Comment:
Can we keep the `autoRenewalDate` property as string and avoid converting it to `Date`?

Except `transferEndDate`, we can just keep the `domain` object untouched and not add or modify any properties to it.

Reply:
We can definitely do that, but that will involve adding date parsing of some sort to every consumer of this selector. And we'll get the asymmetry you mention with `transferEndDate`. Neither of these are huge issues, of course, only trade-offs.

Perhaps I could rename the other properties to `*String` instead of `*Date`, to make it clear the type of data that's expected?

Reply:
As we use moment.js to process the `*Date` properties anyway, I think returning raw date strings from the selector wouldn't be a big issue.

But I'm realizing there's an even better approach: move the `new Date( ... )` code to the [createSiteDomainObject](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/domains/assembler.js#L20) function that converts REST API response to object stored in Redux. Why do a second conversion in selector when it can be done only once in `createSiteDomainObject`?

`trasferEndDate` can be added there, too, and the selector can simply return the `domains` from Redux without any transformation.

We need to check all usages of the `*Date` properties of the `domain` object and make sure the change of type from `String` to `Date` doesn't break anything.

Reply:
Done! I left the ones that didn't have `Date` in the name, even if they were dates, like `expiry`. For those, I kept the string version with the original name, and added a `Date` version, e.g. `expirationDate`, to match what we were already doing in the selector.

Reply:
Looks like I let a failed test slip through. Investigating.

Reply:
Looks like the domains reducer uses a schema, and the schema needs to be updated to match the changes. Unfortunately, as far as I can tell it doesn't support Date objects, only date-formatted strings. Should I just disable schema validation?

Reply:
I _think_ that's because Date objects don't survive the roundtrip to JSON and back during serialization?

Reply:
I'll go ahead and remove the schema, in that case. They're of limited use anyway 🤷‍♂

Reply:
It should also be possible to extend the schema validator with new types.

`Date` objects should be possible to save to IndexedDB (the structure clone algorithm supports them), but I don't know about the Local Storage fallback.

Anyway, it's a good argument to store ISO strings in state and parse them before use in UI code.

Reply:
In that case, I'll go ahead and revert the changes, leaving the transformation in the selector.

Reply:
Done. I moved the `transferEndDate` calculation to the assembler, storing it in state as an ISO date string, and moved date parsing to the consumers.